platform/openide.dialogs/src/org/openide/NotifyDescriptor.java

```diff
@@ -1326,4 +1336,110 @@ public void setSelected(boolean selected) {
             }
         }
     }
+
+    /** Notification providing a composed input of multiple nested selection lists and/or input lines.
+    * @since 7.63
+    */
+    public static final class ComposedInput extends NotifyDescriptor {
+
+        private final List<NotifyDescriptor> inputs = new ArrayList<>();
+        private final Callback callback;
+        private int total;
+
+        /** Construct dialog with the specified title and nested inputs.
+        * @param title title of the dialog
+        * @param total total number of nested inputs
+        * @param callback callback used to create nested inputs
+        * @since 7.63
+        */
+        public ComposedInput(final String title, final int total, final Callback callback) {
+            super(null, title, OK_CANCEL_OPTION, PLAIN_MESSAGE, null, null);
+            this.callback = callback;
+            this.total = total;
+        }
+
+        /**
+         * Total number of nested inputs.
+         * @since 7.63
+         */
+        public int getTotalInputs() {
+            return total;
+        }
+
+        /**
+         * Lazy creates nested input of the given ordinal.
+         * @param number input number from interval &lt;1, totalInputs+1&gt;
+         * @return nested selection list, input line, or null
+         * @since 7.63
+         */
+        public NotifyDescriptor getInput(int number) {
+            NotifyDescriptor step = callback.getInput(number);
+            if (step != null) {
```

Comment:
OK, would it be possible to change the name to `addInput` ? It is somewhat confusing since `getInputs()` returns the current state (array of descriptors) while `getInput(x)` actually creates a `NotifyDescriptor` instance and modifies the state.

Reply:
Perhaps `createInput` could be even better ?